fraud/bad_encoding.go

```diff
@@ -0,0 +1,219 @@
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEnconding{
+		Height:   p.BlockHeight,
+		Shares:   shares,
+		Position: uint32(p.Position),
+		IsRow:    p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
+	in := pb.BadEnconding{}
+	if err := in.Unmarshal(data); err != nil {
+		return err
+	}
+	befp := UnmarshalBefp(&in)
+	*p = *befp
+
+	return nil
+}
+
+func UnmarshalBefp(befp *pb.BadEnconding) *BadEncodingProof {
+	return &BadEncodingProof{
+		BlockHeight: befp.Height,
+		Shares:      ProtoToShare(befp.Shares),
+	}
+}
+
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader, codec rsmt2d.Codec) error {
+	merkleRowRoots := header.DAH.RowsRoots
+	merkleColRoots := header.DAH.ColumnRoots
+	if int(p.Position) >= len(merkleRowRoots) || int(p.Position) >= len(merkleColRoots) {
+		return errors.New("invalid fraud proof: incorrect position of bad row/col")
+	}
+	if len(merkleRowRoots) != len(merkleColRoots) {
+		return errors.New("invalid fraud proof: invalid extended header")
+	}
+	if len(merkleRowRoots) != len(p.Shares) || len(merkleColRoots) != len(p.Shares) {
+		return errors.New("invalid fraud proof: invalid shares")
+	}
+
+	roots := merkleRowRoots
+	if p.isRow {
+		roots = merkleColRoots
+	}
+
+	shares := make([][]byte, len(roots))
+
+	// verify that Merkle proofs correspond to particular shares
+	for index, share := range p.Shares {
+		if share == nil {
+			continue
+		}
+		shares[index] = share.Raw
+		if ok := share.Validate(roots[index]); !ok {
+			return fmt.Errorf("invalid fraud proof: incorrect share received at position %d", index)
+		}
+	}
+
+	// rebuild a row/col
+	rebuiltShares, err := codec.Decode(shares)
+	if err != nil {
+		return err
+	}
+	rebuiltExtendedShares, err := codec.Encode(rebuiltShares)
+	if err != nil {
+		return err
+	}
+	rebuiltShares = append(rebuiltShares, rebuiltExtendedShares...)
+
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(len(shares) / 2))
+	for i, share := range rebuiltShares {
+		tree.Push(share, rsmt2d.SquareIndex{Axis: uint(p.Position), Cell: uint(i)})
+	}
+
+	merkleRoot := merkleColRoots[p.Position]
+	if p.isRow {
+		merkleRoot = merkleRowRoots[p.Position]
+	}
+
+	// compare Merkle Roots
+	if bytes.Equal(tree.Root(), merkleRoot) {
+		return errors.New("invalid fraud proof: merkle root matches")
+	}
+
+	return nil
+}
+
+func buildTreeFromLeaves(
+	ctx context.Context,
+	leaves [][]byte,
+	root []byte,
+	axis int,
+	dag format.NodeGetter,
+) (*wrapper.ErasuredNamespacedMerkleTree, error) {
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(len(leaves) / 2))
+	emptyData := bytes.Repeat([]byte{0}, len(leaves[0]))
+	for idx, data := range leaves {
+		if bytes.Equal(data, emptyData) {
+			// if share was not repaired, then we should get it through sampling
+			share, err := getShare(ctx, root, idx, len(leaves), dag)
```

Comment:
You shouldn't sample the share here. Rather, by construction of the algorithm, this share must have been gotten through sampling before you even tried to repair the square. You just need to save the Merkle proof when you sampled the share originally.

See https://github.com/celestiaorg/celestia-node/issues/263

> The second point will require `retrieveShare` to return a Merkle proof along with the share value.

https://github.com/celestiaorg/celestia-node/blob/40c4c62b7fb0ba8ef26c644c99a741295efd13d8/ipld/read.go#L49